x-pack/plugins/infra/server/lib/host_details/get_services.ts

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { QueryDslQueryContainer } from '@elastic/elasticsearch/lib/api/types';
+import { APMDataAccessConfig } from '@kbn/apm-data-access-plugin/server';
+import { termQuery } from '@kbn/observability-plugin/server';
+import { ESSearchClient } from '../metrics/types';
+import {
+  ServicesAPIRequest,
+  ServicesAPIQueryAggregation,
+} from '../../../common/http_api/host_details';
+import { HOST_NAME_FIELD } from '../../../common/constants';
+
+export const getServices = async (
+  client: ESSearchClient,
+  apmIndices: APMDataAccessConfig['indices'],
+  options: ServicesAPIRequest
+) => {
+  const { error, metric } = apmIndices;
+  const { filters, size, from, to } = options;
+  const commonFiltersList: QueryDslQueryContainer[] = [
+    {
+      range: {
+        '@timestamp': {
+          gte: from,
+          lte: to,
+        },
+      },
+    },
+    {
+      exists: {
+        field: 'service.name',
+      },
+    },
+  ];
+
+  if (filters['host.name']) {
+    // also query for host.hostname field along with host.name, as some services may use this field
+    const HOST_HOSTNAME_FIELD = 'host.hostname';
+    commonFiltersList.push({
+      bool: {
+        should: [
+          ...termQuery(HOST_NAME_FIELD, filters[HOST_NAME_FIELD]),
+          ...termQuery(HOST_HOSTNAME_FIELD, filters[HOST_NAME_FIELD]),
+        ],
+        minimum_should_match: 1,
+      },
+    });
+  }
+  const aggs = {
+    services: {
+      terms: {
+        field: 'service.name',
+        size,
+      },
+      aggs: {
+        latestAgent: {
+          top_metrics: {
+            metrics: [{ field: 'agent.name' }],
+            sort: {
+              '@timestamp': 'desc',
+            },
+            size: 1,
+          },
+        },
+      },
+    },
+  };
+  // get services from transaction metrics
+  const metricsQuery = {
+    size: 0,
+    _source: false,
+    query: {
+      bool: {
+        filter: [
+          {
+            term: {
+              'metricset.name': 'transaction',
+            },
+          },
+          ...commonFiltersList,
+        ],
```

Comment:
You should also specify `metricset.interval`. Possible values are 1m, 10m and 60m. You should pick the interval based on the selected time range. If the time range duration is 1 hour you should normally use 1m interval for time series - or possibly 10m interval if it's a single metric (number). 

If the time range duration is more than 24h you should always use 60m interval

Reply:
To keep things simple as I don't think it will much of a performance impact, I've set it to 1m in all cases.  We don't currently have a requirement to get any timeseries data.  The default timerange is last 15 minutes.

Reply:
> To keep things simple 

I 100% agree with this - which is why I have been arguing against the OAM approach where we try to generalize and re-use. As long as this is not seen as some part of OAM which will be re-used elsewhere I think it's fine for now to take a shortcut and simplify.